core/src/main/java/org/apache/iceberg/SnapshotManager.java

```diff
@@ -50,25 +51,34 @@ protected ManageSnapshots self() {
 
   @Override
   protected String operation() {
-    // snapshotOperation is used by SnapshotProducer when building and writing a new snapshot for cherrypick
-    Preconditions.checkNotNull(snapshotOperation, "[BUG] Detected uninitialized operation");
+    //TODO: We are not setting this var during rollbacks. What should its value be during rollbacks?
     return snapshotOperation;
   }
 
+  @Override
+  protected long snapshotId() {
+    if (requiresSnapshotCreation) {
+      return super.snapshotId();
+    }
+    // No new snapshot was created
+    return targetSnapshotId;
+  }
+
   @Override
   public ManageSnapshots cherrypick(long snapshotId) {
     TableMetadata current = current();
     ValidationException.check(current.snapshot(snapshotId) != null,
         "Cannot cherry pick unknown snapshot id: %s", snapshotId);
 
     Snapshot cherryPickSnapshot = current.snapshot(snapshotId);
+    // TODO: shouldn't we check for appends only if cherrypick op is not fast forward-able?
```

Comment:
No, the problem is that it may look at configuration time like the operation will be a fast-forward, but concurrent writes might require a cherry-pick instead. We could handle that by failing in apply, but it's okay to make sure the commit _could_ be cherry-picked if needed.